src/common/gpio_motor.rs

```diff
@@ -0,0 +1,404 @@
+        }
+        Ok(None)
+    }
+}
+
+impl<B> Status for ABPwmMotor<B>
+where
+    B: Board,
+{
+    fn get_status(&self) -> anyhow::Result<Option<google::protobuf::Struct>> {
+        let mut hm = HashMap::new();
+        let pos = 0.0;
+        hm.insert(
+            "position".to_string(),
+            google::protobuf::Value {
+                kind: Some(google::protobuf::value::Kind::NumberValue(pos)),
+            },
+        );
+        Ok(Some(google::protobuf::Struct { fields: hm }))
+    }
+}
+
+impl<B> Stoppable for ABPwmMotor<B>
+where
+    B: Board,
+{
+    fn stop(&mut self) -> anyhow::Result<()> {
+        self.set_power(0.0)
+    }
+}
+
+// Represents a motor using a direction pin and a PWM pin
+pub struct PwmDirectionMotor<B> {
+    board: B,
+    dir_pin: i32,
+    pwm_pin: i32,
+    max_rpm: f64,
+    dir_flip: bool,
+}
+
+impl<B> PwmDirectionMotor<B>
+where
+    B: Board,
+{
+    fn new(
+        dir_pin: i32,
+        pwm_pin: i32,
+        max_rpm: f64,
+        dir_flip: bool,
+        board: B,
+    ) -> anyhow::Result<Self> {
+        let mut res = Self {
+            board,
+            dir_pin,
+            pwm_pin,
+            max_rpm,
+            dir_flip,
+        };
+        res.board.set_pwm_frequency(pwm_pin, 10000)?;
+        Ok(res)
+    }
+
+    pub(crate) fn from_config(cfg: ConfigType, board: BoardType) -> anyhow::Result<MotorType> {
+        if let Ok(pins) = cfg.get_attribute::<MotorPinsConfig>("pins") {
+            if let Some(dir_pin) = pins.dir {
+                let pwm_pin = pins.pwm;
+                let max_rpm: f64 = cfg.get_attribute::<f64>("max_rpm")?;
+                let dir_flip: bool = cfg.get_attribute::<bool>("dir_flip").unwrap_or_default();
+                return Ok(Arc::new(Mutex::new(PwmDirectionMotor::new(
+                    dir_pin, pwm_pin, max_rpm, dir_flip, board,
+                )?)));
+            }
+        }
+        Err(anyhow::anyhow!("cannot build motor"))
+    }
+}
+
+impl<B> Motor for PwmDirectionMotor<B>
+where
+    B: Board,
+{
+    fn set_power(&mut self, pct: f64) -> anyhow::Result<()> {
+        if !(-1.0..=1.0).contains(&pct) {
+            anyhow::bail!("power outside limit")
+        }
+        let set_high = (pct > 0.0) && !self.dir_flip;
+        self.board.set_gpio_pin_level(self.dir_pin, set_high)?;
+        self.board.set_pwm_duty(self.pwm_pin, pct)?;
+        Ok(())
+    }
+
+    fn get_position(&mut self) -> anyhow::Result<i32> {
+        Ok(0)
+    }
+
+    fn go_for(&mut self, rpm: f64, revolutions: f64) -> anyhow::Result<Option<Duration>> {
+        let (pwr, dur) = go_for_math(self.max_rpm, rpm, revolutions)?;
+        self.set_power(pwr)?;
+        if dur.is_some() {
+            return Ok(dur);
+        }
+        Ok(None)
+    }
+}
+
+impl<B> Status for PwmDirectionMotor<B>
+where
+    B: Board,
+{
+    fn get_status(&self) -> anyhow::Result<Option<google::protobuf::Struct>> {
+        let mut hm = HashMap::new();
+        let pos = 0.0;
```

Comment:
Just a question, but why would an unencoded motor report a position in its status?

Reply:
In order to remove the position from this map, we have to implement `GetProperties` in the motor API because otherwise the position badge on the motor RC card will read `NaN`. Should I make a ticket for implementing `GetProperties`?

Reply:
ticket created: https://viam.atlassian.net/browse/RSDK-5331